cmd/faFindFast/efficient.go

```diff
@@ -4,133 +4,210 @@ import (
-					if totalSubst+totalGaps > windowSize {
-						log.Fatalf("Error: total number of mutations exceeds windowSize. This may or may not be a bug, but your sequence has deviated from our use case.")
-					}
-					rawPValue = scorePValueCache[totalSubst+totalGaps]
-					_, err = fmt.Fprintf(out, "%s\t%d\t%d\t%s_%d\t%d\t%s\t%e\t%e\n", refChrName, refIdxBeforeWindow+1, lastRefIdxOfWindow+1, refChrName, refIdxBeforeWindow+1, totalSubst+totalGaps, "+", percentDiverged, rawPValue)
-					exception.PanicOnErr(err)
+		if lastFirstQueryIdxOfWindow-firstQueryIdxBeforeWindow == s.WindowSize && lastAlnIdxOfWindow < len(firstQuery) {
+
+			// convert position from firstQuery to reference
+			// check if firstQuery chromStart (non-gap) is gap in reference. If so, don't report that window
+			// check if firstQuery chromEnd-1 (last non-gap position) is gap in reference. If so, don't report that window
+			if !((reference[alnIdxBeforeWindowForRef+1] == dna.Gap && firstQuery[alnIdxBeforeWindowForRef+1] != dna.Gap) || (reference[lastAlnIdxOfWindow] == dna.Gap && firstQuery[lastAlnIdxOfWindow] != dna.Gap)) {
+				// call fasta.AlnPosToRefPosCounterSeq on the alnPos of chromStart and chromEnd
+				// as opposed to alnIdxBeforeWindowForRef and lastAlnIdxOfWindow
+				// this is because fasta.AlnPosToRefPosCounterSeq rounds position up when it encounters gap
+				// in order to only scan the genome once, call fasta.AlnPosToRefPosCounterSeq on saved refStart and alnStart once refStart and alnStart both exceed 0
+				if lastRefIdxOfWindowPlusOne < 0 || lastAlnIdxOfWindow+1 < 0 {
+					refIdxWindowStart = fasta.AlnPosToRefPosCounterSeq(reference, alnIdxBeforeWindowForRef+1, 0, 0)
 				} else {
-					_, err = fmt.Fprintf(out, "%s\t%d\t%d\t%s_%d\t%d\n", refChrName, refIdxBeforeWindow+1, lastRefIdxOfWindow+1, refChrName, refIdxBeforeWindow+1, totalSubst+totalGaps)
-					exception.PanicOnErr(err)
+					refIdxWindowStart = fasta.AlnPosToRefPosCounterSeq(reference, alnIdxBeforeWindowForRef+1, lastRefIdxOfWindowPlusOne, lastAlnIdxOfWindow+1)
 				}
-			}
+				lastRefIdxOfWindowPlusOne = fasta.AlnPosToRefPosCounterSeq(reference, lastAlnIdxOfWindow+1, refIdxWindowStart, alnIdxBeforeWindowForRef+1)
+
+				// an option/flag can tell us not to print if there are Ns in the firstQuery or secondQuery
+				if !s.RemoveN || totalNs == 0 {
+					if s.LongOutput && !s.OutputAlnPos {
+						percentDiverged = 100 * (float64(totalSubst+totalGaps) / float64(s.WindowSize))
+						if totalSubst+totalGaps > s.WindowSize {
+							log.Fatalf("Error: total number of mutations exceeds windowSize. This may or may not be a bug, but your sequence has deviated from our use case.\n")
+						}
+						rawPValue = scorePValueCache[totalSubst+totalGaps]
+						_, err = fmt.Fprintf(file, "%s\t%d\t%d\t%s_%d\t%d\t%s\t%e\t%e\n", s.RefChromName, refIdxWindowStart, lastRefIdxOfWindowPlusOne, s.RefChromName, refIdxWindowStart, totalSubst+totalGaps, "+", percentDiverged, rawPValue)
```

Comment:
in old code would print (start, end) as (refIdxBeforeWindow+1, lastRefIdxOfWindow+1). note that ref in old code is firstQuery in new code. Now, directly use fasta.AlnPosToRefPosCounterSeq to calculate the bed entry's (start, end) as (refIdxWindowStart, lastRefIdxOfWindowPlusOne). This is because if we convert from firstQueryBeforeWindow to refIdxWindowStart-1, we can't guarantee that firstQueryBeforeWindow+1 converts to refIdxWindowStart. Smilarly, if we convert from lastRefIdxOfWindow to lastRefIdxOfWindow, we can't guarantee that lastRefIdxOfWindow+1 converts to lastRefIdxOfWindowPlusOne